torch/_inductor/runtime/runtime_utils.py

```diff
@@ -81,35 +80,62 @@ def do_bench(fn, fn_args, fn_kwargs, **kwargs):
         return do_bench_gpu(lambda: fn(*fn_args, **fn_kwargs), **kwargs)
 
 
-def do_bench_gpu(*args, **kwargs):
+def do_bench_gpu(fn, warmup=25, rep=100, fast_flush=True):
     @functools.lru_cache(None)
-    def load_triton():
-        try:
-            # NB: Lazily load triton, as importing triton is slow
-            # see https://github.com/openai/triton/issues/1599
-            from triton.testing import do_bench as triton_do_bench
-        except ImportError as exc:
-            raise NotImplementedError("requires Triton") from exc
-
-        # triton PR https://github.com/openai/triton/pull/1513 change the
-        # quantile fields name from 'percentiles' to 'quantiles'
-        # and change the default value from (0.5, 0.2, 0.8) to None.
-        # This may break inductor since a caller expects a tuple may get a item.
-        #
-        # Add a wrapper to maintain the same behavior for inductor.
-        # Maybe we should have own implementation of this function?
-        return triton_do_bench, (
-            "quantiles"
-            if inspect.signature(triton_do_bench).parameters.get("quantiles")
-            is not None
-            else "percentiles"
+    def get_cache_size():
+        device = torch.cuda.current_device()
+        properties = torch.cuda.get_device_properties(device)
+        return properties.l2CacheSize
+
+    if fast_flush:
+        cache = torch.empty(int(get_cache_size() // 4), dtype=torch.int, device="cuda")
+    else:
+        cache = torch.empty(int(256e6), dtype=torch.int8, device="cuda")
+
+    # Estimate the runtime of the function
+    estimation_iters = 5
+    event_pairs = [
+        (
+            torch.cuda.Event(enable_timing=True),
+            torch.cuda.Event(enable_timing=True),
         )
+        for _ in range(estimation_iters)
+    ]
+    for start_event, end_event in event_pairs:
+        cache.zero_()
+        start_event.record()
+        fn()
+        end_event.record()
+    torch.cuda.synchronize()
+    estimate_ms = min(
+        [event_pair[0].elapsed_time(event_pair[1]) for event_pair in event_pairs]
+    )
+
+    warmup_iters = min(10, max(1, int(warmup / estimate_ms)))
+    repeat_iters = min(40, max(1, int(rep / estimate_ms)))
 
-    triton_do_bench, quantile_field_name = load_triton()
+    event_pairs = [
+        (
+            torch.cuda.Event(enable_timing=True),
+            torch.cuda.Event(enable_timing=True),
+        )
+        for _ in range(repeat_iters)
+    ]
 
-    if quantile_field_name not in kwargs:
-        kwargs[quantile_field_name] = (0.5, 0.2, 0.8)
-    return triton_do_bench(*args, **kwargs)[0]
+    for _ in range(warmup_iters):
+        fn()
+
+    for start_event, end_event in event_pairs:
+        cache.zero_()
+        start_event.record()
+        fn()
+        end_event.record()
+    torch.cuda.synchronize()
+
+    timing = min(
+        [event_pair[0].elapsed_time(event_pair[1]) for event_pair in event_pairs]
+    )
```

Comment:
From https://github.com/triton-lang/triton/pull/3943#issuecomment-2128027979: 

> I agree that min is much more stable than mean, and that's probably what I'd do if I were starting from scratch. But this biases you towards algorithms which perform better when the GPU is cold, and I believe we have seen cases where that isn't globally ideal. Just adding to the complexity...

Could we maintain current behavior of median and consider switching in a future pr ?

Or continue to expose `return_mode`

Reply:
Sure, good idea to seperate that out into a future PR.

More on that idea though, I think both min and quantile are inherently flawed in their own ways. This all loops back to the main issue that autotuning is really hard, and you can find problems which pretty much every way of doing it. 

Yes, min is can have problems when running on a cold GPU. But really, isn't that what the warmups are supposed to take care of? How many warmups are enough to get the GPU warm without cooking it? It is very hard to get this even close to right.

Any sort of quantile also has its own problems. The quantiles can get pretty heavily skewed if anything disturbs the GPU for a portion of the benchmarking; min is not as sensitive to this issue.

I'll split this off into a separate PR like you mentioned; it really has no bearing on compile time overhead, and is therefore not so relevant to this PR.